messages/src/lib.rs

```diff
@@ -41,6 +43,80 @@ pub enum Error {
     UnsupportedAlgorithmIdentifier(&'static str, u16),
 }
 
+/// Wire-representation of an ASCII-encoded URL with minimum length 1 and maximum
+/// length 2^16 - 1. This is a newtype for [`url::Url`] for adding encoding and
+/// decoding traits.
+///
+/// Note that because [`url::Url::parse`] is used while decoding, this type has more
+/// restrictions on it than specified in DAP. See [`url::ParseError`] for the additional
+/// ways decoding can fail.
+///
+/// Also note that the encoding of a URL may not precisely match what it was decoded
+/// from. In particular, the in-memory representation of [`url::Url`] adds a trailing
+/// slash if it is appropriate to do so, e.g. `https://example.com -> https://example.com/`.
+/// This does not affect the parseability of the resulting encoded Url, but may result
+/// in an increase in size of the encoding.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Url(url::Url);
+
+impl Url {
+    const MAX_LEN: usize = 2usize.pow(16) - 1;
+
+    // Access the underlying [`url::Url`].
+    pub fn url(&self) -> &url::Url {
+        &self.0
+    }
```

Comment:
I was heavily tempted to implement `Deref` for this, but the doc says thou shalt only implement this for smart pointers. https://doc.rust-lang.org/std/ops/trait.Deref.html

And opinion seems all over the place for implementing `Deref` on [newtypes anyways](https://stackoverflow.com/questions/45086595/is-it-considered-a-bad-practice-to-implement-deref-for-newtypes). Do we have any preference?

Reply:
I don't have a strong preference. I'm OK following the docs' suggestion here.